controllers/hostdefiner_controller.go

```diff
@@ -0,0 +1,453 @@
+	logger := hostDefinerLog.WithValues("Resource Type", "ServiceAccount")
+
+	hostDefiner := instance.GenerateServiceAccount()
+
+	for _, sa := range []*corev1.ServiceAccount{
+		hostDefiner,
+	} {
+		if err := controllerutil.SetControllerReference(instance.Unwrap(), sa, r.Scheme); err != nil {
+			return err
+		}
+		found := &corev1.ServiceAccount{}
+		err := r.Get(context.TODO(), types.NamespacedName{
+			Name:      sa.Name,
+			Namespace: sa.Namespace,
+		}, found)
+		if err != nil && errors.IsNotFound(err) {
+			logger.Info("Creating a new ServiceAccount", "Namespace", sa.GetNamespace(), "Name", sa.GetName())
+			err = r.Create(context.TODO(), sa)
+			if err != nil {
+				return err
+			}
+
+			rErr := r.restartDeployment(logger, instance)
+			if rErr != nil {
+				return rErr
+			}
+
+		} else if err != nil {
+			logger.Error(err, "Failed to get ServiceAccount", "Name", sa.GetName())
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *HostDefinerReconciler) restartDeployment(logger logr.Logger, instance *hostdefiner.HostDefiner) error {
+	deployment, err := r.getDeployment(instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	logger.Info("hostDefiner requires restart",
+		"ReadyReplicas", deployment.Status.ReadyReplicas,
+		"Replicas", deployment.Status.Replicas)
+	logger.Info("restarting csi hostDefiner")
+
+	err = r.rolloutRestartDeployment(deployment)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *HostDefinerReconciler) getDeployment(instance *hostdefiner.HostDefiner) (*appsv1.Deployment, error) {
+	deployment := &appsv1.Deployment{}
+	err := r.Get(context.TODO(), types.NamespacedName{
+		Name:      oconfig.GetNameForResource(oconfig.CSIHostDefiner, instance.Name),
+		Namespace: instance.Namespace,
+	}, deployment)
+
+	return deployment, err
+}
+
+func (r *HostDefinerReconciler) rolloutRestartDeployment(deployment *appsv1.Deployment) error {
+	restartedAt := fmt.Sprintf("%s/restartedAt", oconfig.APIGroup)
+	timestamp := time.Now().String()
+	deployment.Spec.Template.ObjectMeta.Annotations[restartedAt] = timestamp
+	return r.Update(context.TODO(), deployment)
+}
+
+func (r *HostDefinerReconciler) updateStatus(instance *hostdefiner.HostDefiner, originalStatus csiv1.HostDefinerStatus) error {
+	logger := log.WithName("updateStatus")
+	deployment, err := r.getDeployment(instance)
+	if err != nil {
+		return err
+	}
+
+	r.updateStatusFields(instance, deployment)
+
+	if !reflect.DeepEqual(originalStatus, instance.Status) {
+		logger.Info("updating HostDefiner status", "name", instance.Name, "from", originalStatus, "to", instance.Status)
+		sErr := r.Status().Update(context.TODO(), instance.Unwrap())
+		if sErr != nil {
+			return sErr
+		}
+	}
+
+	return nil
+}
+
+func (r *HostDefinerReconciler) updateStatusFields(instance *hostdefiner.HostDefiner, deployment *appsv1.Deployment) {
+	instance.Status.HostDefinerReady = r.isReady(deployment)
+	phase := csiv1.DriverPhaseCreating
```

Comment:
I don't think we should use "Driver"

Reply:
I wanted to use the same variables as the csi controller because they need the same values, so you think I should change all the [phase](https://github.com/IBM/ibm-block-csi-operator/blob/0e1a2a144d2daea840075f58945181c1a729a0a4/api/v1/ibmblockcsi_types.go#L26-L30) variables to more common name and move it to common config file?

Reply:
sounds good to me

Reply:
I don't see any change here

Reply:
same for every place that refers to the host definer as the driver (e.g. [here](https://github.com/IBM/ibm-block-csi-operator/blob/2b34f83399f8e0dc9d3b5977cac5c2d67c13dcb6/controllers/syncer/csi_hostdefiner.go#L42))

Reply:
done - #271